File: src/rsmod/LinePathFinder.ts
import CollisionFlagMap from './collision/CollisionFlagMap';
import Line from './Line';
import {CollisionFlag} from './flag/CollisionFlag';
import RouteCoordinates from './RouteCoordinates';

@final
export default class LinePathFinder {
    private static readonly EMPTY: StaticArray<i32> = new StaticArray(0);

    private readonly flags: CollisionFlagMap;

    constructor(flags: CollisionFlagMap) {
        this.flags = flags;
    }

    // prettier-ignore
    @inline
    lineOfSight(
        level: i32,
        srcX: i32,
        srcZ: i32,
        destX: i32,
        destZ: i32,
        srcWidth: i32,
        srcHeight: i32,
        destWidth: i32,
        destHeight: i32,
        extraFlag: i32
    ): StaticArray<i32> {
        return this.rayCast(
            level,
            srcX,
            srcZ,
            destX,
            destZ,
            srcWidth,
            srcHeight,
            destWidth,
            destHeight,
            Line.SIGHT_BLOCKED_WEST | extraFlag,
            Line.SIGHT_BLOCKED_EAST | extraFlag,
            Line.SIGHT_BLOCKED_SOUTH | extraFlag,
            Line.SIGHT_BLOCKED_NORTH | extraFlag,
            CollisionFlag.LOC | extraFlag,
            CollisionFlag.LOC_PROJ_BLOCKER | extraFlag,
            true
        );
    }

    // prettier-ignore
    @inline
    lineOfWalk(
        level: i32,
        srcX: i32,
        srcZ: i32,
        destX: i32,
        destZ: i32,
        srcWidth: i32,
        srcHeight: i32,
        destWidth: i32,
        destHeight: i32,
        extraFlag: i32
    ): StaticArray<i32> {
        return this.rayCast(
            level,
            srcX,
            srcZ,
            destX,
            destZ,
            srcWidth,
            srcHeight,
            destWidth,
            destHeight,
            Line.WALK_BLOCKED_WEST | extraFlag,
            Line.WALK_BLOCKED_EAST | extraFlag,
            Line.WALK_BLOCKED_SOUTH | extraFlag,
            Line.WALK_BLOCKED_NORTH | extraFlag,
            CollisionFlag.LOC | extraFlag,
            CollisionFlag.LOC_PROJ_BLOCKER | extraFlag,
            false
        );
    }

    // prettier-ignore
    @inline
    private rayCast(
        level: i32,
        srcX: i32,
        srcZ: i32,
        destX: i32,
        destZ: i32,
        srcWidth: i32,
        srcHeight: i32,
        destWidth: i32,
        destHeight: i32,
        flagWest: i32,
        flagEast: i32,
        flagSouth: i32,
        flagNorth: i32,
        flagLoc: i32,
        flagProj: i32,
        los: bool
    ): StaticArray<i32> {
        const startX: i32 = Line.coordinate(srcX, destX, srcWidth);
        const startZ: i32 = Line.coordinate(srcZ, destZ, srcHeight);

        const endX: i32 = Line.coordinate(destX, srcX, destWidth);
        const endZ: i32 = Line.coordinate(destZ, srcZ, destHeight);

        if (startX == endX && startZ == endZ) {
            return LinePathFinder.EMPTY;
        }

        if (los && this.flags.isFlagged(startX, startZ, level, flagLoc)) {
            return LinePathFinder.EMPTY;
        }

        const deltaX: i32 = endX - startX;
        const deltaZ: i32 = endZ - startZ;
        const absoluteDeltaX: i32 = <i32>Math.abs(deltaX);
        const absoluteDeltaZ: i32 = <i32>Math.abs(deltaZ);

        const travelEast: bool = deltaX >= 0;
        const travelNorth: bool = deltaZ >= 0;

        let xFlags: i32 = travelEast ? flagWest : flagEast;
        let zFlags: i32 = travelNorth ? flagSouth : flagNorth;

        const coordinates: i32[] = [];
        if (absoluteDeltaX > absoluteDeltaZ) {
            const offsetX: i32 = travelEast ? 1 : -1;
            const offsetZ: i32 = travelNorth ? 0 : -1;

            let scaledZ: i32 = Line.scaleUp(startZ) + Line.HALF_TILE + offsetZ;
            const tangent: i32 = Line.scaleUp(deltaZ) / absoluteDeltaX;

            let currX: i32 = startX;
            while (currX != endX) {
                currX += offsetX;
                const currZ: i32 = Line.scaleDown(scaledZ);
                if (los && currX == endX && currZ == endZ) {
                    xFlags = xFlags & ~flagProj;
                }
                if (this.flags.isFlagged(currX, currZ, level, xFlags)) {
                    return LinePathFinder.EMPTY; // alternative
                }
                coordinates.push(RouteCoordinates.pack(level, currX, currZ));

                scaledZ += tangent;

                const nextZ: i32 = Line.scaleDown(scaledZ);
                if (nextZ != currZ) {
                    if (los && currX == endX && nextZ == endZ) {
                        zFlags = zFlags & ~flagProj;
                    }
                    if (this.flags.isFlagged(currX, nextZ, level, zFlags)) {
                        return LinePathFinder.EMPTY; // alternative
                    }
                    coordinates.push(RouteCoordinates.pack(level, currX, nextZ));
                }
            }
        } else {
            const offsetX: i32 = travelEast ? 0 : -1;
            const offsetZ: i32 = travelNorth ? 1 : -1;

            let scaledX: i32 = Line.scaleUp(startX) + Line.HALF_TILE + offsetX;
            const tangent: i32 = Line.scaleUp(deltaX) / absoluteDeltaZ;

            let currZ: i32 = startZ;
            while (currZ != endZ) {
                currZ += offsetZ;
                const currX: i32 = Line.scaleDown(scaledX);
                if (los && currX == endX && currZ == endZ) {
                    zFlags = zFlags & ~flagProj;
                }
                if (this.flags.isFlagged(currX, currZ, level, zFlags)) {
                    return LinePathFinder.EMPTY; // alternative
                }
                coordinates.push(RouteCoordinates.pack(level, currX, currZ));

                scaledX += tangent;

                const nextX: i32 = Line.scaleDown(scaledX);
                if (nextX != currX) {
                    if (los && nextX == endX && currZ == endZ) {
                        xFlags = xFlags & ~flagProj;
                    }
                    if (this.flags.isFlagged(nextX, currZ, level, xFlags)) {
                        return LinePathFinder.EMPTY; // alternative
                    }
                    coordinates.push(RouteCoordinates.pack(level, nextX, currZ));
                }
            }
        }
        const route: StaticArray<i32> = new StaticArray<i32>(coordinates.length);
        for (let i: i32 = 0; i < coordinates.length; i++) {
            unchecked(route[i] = coordinates[i]);
        }
        return route;
    }
}

File: src/rsmod/line_pathfinder.rs
use crate::rsmod::collision::collision::CollisionFlagMap;
use crate::rsmod::collision_flag::CollisionFlag;
use crate::rsmod::CoordGrid;
use crate::rsmod::line::Line;

#[inline(always)]
pub unsafe fn line_of_sight(
    flags: &CollisionFlagMap,
    y: i32,
    src_x: i32,
    src_z: i32,
    dest_x: i32,
    dest_z: i32,
    src_width: u8,
    src_height: u8,
    dest_width: u8,
    dest_height: u8,
    extra_flag: u32,
) -> Vec<u32> {
    return rayCastPath(
        flags,
        y,
        src_x,
        src_z,
        dest_x,
        dest_z,
        src_width,
        src_height,
        dest_width,
        dest_height,
        Line::SIGHT_BLOCKED_WEST | extra_flag,
        Line::SIGHT_BLOCKED_EAST | extra_flag,
        Line::SIGHT_BLOCKED_SOUTH | extra_flag,
        Line::SIGHT_BLOCKED_NORTH | extra_flag,
        CollisionFlag::LOC as u32 | extra_flag,
        CollisionFlag::LOC_PROJ_BLOCKER as u32 | extra_flag,
        true,
    );
}

#[inline(always)]
pub unsafe fn line_of_walk(
    flags: &CollisionFlagMap,
    y: i32,
    src_x: i32,
    src_z: i32,
    dest_x: i32,
    dest_z: i32,
    src_width: u8,
    src_height: u8,
    dest_width: u8,
    dest_height: u8,
    extra_flag: u32,
) -> Vec<u32> {
    return rayCastPath(
        flags,
        y,
        src_x,
        src_z,
        dest_x,
        dest_z,
        src_width,
        src_height,
        dest_width,
        dest_height,
        Line::WALK_BLOCKED_WEST | extra_flag,
        Line::WALK_BLOCKED_EAST | extra_flag,
        Line::WALK_BLOCKED_SOUTH | extra_flag,
        Line::WALK_BLOCKED_NORTH | extra_flag,
        CollisionFlag::LOC as u32 | extra_flag,
        CollisionFlag::LOC_PROJ_BLOCKER as u32 | extra_flag,
        false,
    );
}

#[inline(always)]
unsafe fn rayCastPath(
    flags: &CollisionFlagMap,
    y: i32,
    src_x: i32,
    src_z: i32,
    dest_x: i32,
    dest_z: i32,
    src_width: u8,
    src_height: u8,
    dest_width: u8,
    dest_height: u8,
    flag_west: u32,
    flag_east: u32,
    flag_south: u32,
    flag_north: u32,
    flag_loc: u32,
    flag_proj: u32,
    los: bool,
) -> Vec<u32> {
    let start_x: i32 = Line::coordinate(src_x, dest_x, src_width);
    let start_z: i32 = Line::coordinate(src_z, dest_z, src_height);

    let end_x: i32 = Line::coordinate(dest_x, src_x, dest_width);
    let end_z: i32 = Line::coordinate(dest_z, src_z, dest_height);

    if start_x == end_x && start_z == end_z {
        return vec![];
    }

    if los && flags.is_flagged(start_x, start_z, y, flag_loc) {
        return vec![];
    }

    let delta_x: i32 = end_x - start_x;
    let delta_z: i32 = end_z - start_z;
    let absolute_delta_x: i32 = delta_x.abs();
    let absolute_delta_z: i32 = delta_z.abs();

    let travel_east: bool = delta_x >= 0;
    let travel_north: bool = delta_z >= 0;

    let mut x_flags: u32 = if travel_east { flag_west } else { flag_east };
    let mut z_flags: u32 = if travel_north { flag_south } else { flag_north };

    let mut coordinates: Vec<u32> = vec![];
    if absolute_delta_x > absolute_delta_z {
        let offset_x: i32 = if travel_east { 1 } else { -1 };
        let offset_z: i32 = if travel_north { 0 } else { -1 };

        let mut scaled_z: i32 = Line::scale_up(start_z) + Line::HALF_TILE + offset_z;
        let tangent: i32 = Line::scale_up(delta_z) / absolute_delta_x;

        let mut curr_x: i32 = start_x;
        while curr_x != end_x {
            curr_x += offset_x;
            let curr_z: i32 = Line::scale_down(scaled_z);
            if los && curr_x == end_x && curr_z == end_z {
                x_flags &= !flag_proj;
            }
            if flags.is_flagged(curr_x, curr_z, y, x_flags) {
                return vec![]; // alternative
            }
            coordinates.push(CoordGrid::new(y, curr_x, curr_z).packed);

            scaled_z += tangent;

            let nextZ: i32 = Line::scale_down(scaled_z);
            if nextZ != curr_z {
                if los && curr_x == end_x && nextZ == end_z {
                    z_flags &= !flag_proj;
                }
                if flags.is_flagged(curr_x, nextZ, y, z_flags) {
                    return vec![]; // alternative
                }
                coordinates.push(CoordGrid::new(y, curr_x, nextZ).packed);
            }
        }
    } else {
        let offset_x: i32 = if travel_east { 0 } else { -1 };
        let offset_z: i32 = if travel_north { 1 } else { -1 };

        let mut scaled_x: i32 = Line::scale_up(start_x) + Line::HALF_TILE + offset_x;
        let tangent: i32 = Line::scale_up(delta_x) / absolute_delta_z;

        let mut curr_z: i32 = start_z;
        while curr_z != end_z {
            curr_z += offset_z;
            let curr_x: i32 = Line::scale_down(scaled_x);
            if los && curr_x == end_x && curr_z == end_z {
                z_flags &= !flag_proj;
            }
            if flags.is_flagged(curr_x, curr_z, y, z_flags) {
                return vec![]; // alternative
            }
            coordinates.push(CoordGrid::new(y, curr_x, curr_z).packed);

            scaled_x += tangent;

            let nextX: i32 = Line::scale_down(scaled_x);
            if nextX != curr_x {
                if los && nextX == end_x && curr_z == end_z {
                    x_flags &= !flag_proj;
                }
                if flags.is_flagged(nextX, curr_z, y, x_flags) {
                    return vec![]; // alternative
                }
                coordinates.push(CoordGrid::new(y, nextX, curr_z).packed);
            }
        }
    }
    return coordinates;
}
